### PR TITLE
KeyringController: improve unit test coverage

### PR DESCRIPTION
**Pull request by jpuri**
Adding more unit test cases for keyring controller.

### CODE REVIEW

File: src/keyring/KeyringController.test.ts [hunk 1]
```diff
@@ -247,6 +346,36 @@ describe('KeyringController', () => {
     expect(account).toBe(recovered);
   });
```
Comment (jpuri) [inline src/keyring/KeyringController.test.ts:348]: For method below `signPersonalMessage` it seems not to fail for empty data, I am not sure if that is ok.

File: src/keyring/KeyringController.test.ts [hunk 2]
```diff
@@ -459,11 +616,64 @@ describe('KeyringController', () => {
...
+      await keyringController.signTransaction(unsignedEthTx, '');
+    } catch (e) {
+      error = e;
+    }
+
+    expect(error.message).toContain(
+      'No keyring found for the requested account. Error info: The address passed in is invalid/empty; There are keyrings, but none match the address;',
+    );
+  });
+
```
Comment (jpuri) [inline src/keyring/KeyringController.test.ts:646]: `signTransaction` should fail for incorrect transaction, currently the check is not great

File: src/keyring/KeyringController.test.ts [hunk 3]
```diff
@@ -459,11 +616,64 @@ describe('KeyringController', () => {
...
+      error = e;
+    }
+
+    expect(error.message).toContain('tx.sign is not a function');
+  });
+
   it('should submit password and decrypt', async () => {
     const state = await keyringController.submitPassword(password);
     expect(state).toStrictEqual(initialState);
   });
```
Comment (jpuri) [inline src/keyring/KeyringController.test.ts:668]: `keyringController.submitPassword` returns current state even if wrong password is passed. This probably needs fix.

File: src/keyring/KeyringController.test.ts [hunk 4]
```diff
@@ -119,7 +123,29 @@ describe('KeyringController', () => {
...
+      password,
+      currentSeedWord,
+    );
+    expect(initialState).toStrictEqual(currentState);
+  });
+
+  it('should throw error if creating new vault and restore without password', async () => {
+    await expect(async () => {
+      await keyringController.createNewVaultAndRestore('', seedWords);
+    }).rejects.toThrow('Invalid password');
```
Comment (mcmire) [inline src/keyring/KeyringController.test.ts:138]: [nit] `expect` also takes a promise directly, so if you like, you don't need the function here, you can shorten this to:

``` suggestion
    await expect(
      keyringController.createNewVaultAndRestore('', seedWords)
    ).rejects.toThrow('Invalid password');
```

That said, I see this pattern being used frequently in this file, so not a big deal — just a shortcut going forward.

File: src/keyring/KeyringController.test.ts [hunk 5]
```diff
@@ -449,8 +563,42 @@ describe('KeyringController', () => {
     } catch (e) {
       error2 = e;
     }
-    expect(error1.message).toContain('Keyring Controller signTypedMessage:');
-    expect(error2.message).toContain('Keyring Controller signTypedMessage:');
+
+    expect((error1 as ErrorMessageObject).message).toContain(
```
Comment (mcmire) [inline src/keyring/KeyringController.test.ts:567]: Instead of checking for the error like this, why not say:

``` javascript
expect(
  keyringController.signTypedMessage(
    { data: msgParams, from: account },
    SignTypedDataVersion.V1,
  )
).rejects.toThrow(/^Keyring Controller signTypedMessage:/u)
```

This way you don't have to typecast the error.

File: src/keyring/KeyringController.test.ts [hunk 6]
```diff
@@ -477,11 +625,54 @@ describe('KeyringController', () => {
...
+
   it('should submit password and decrypt', async () => {
     const state = await keyringController.submitPassword(password);
     expect(state).toStrictEqual(initialState);
   });
 
+  /**
+   * TODO: test below should fail for wrong password
+   */
+  it('should fail on submit password for wrong password', async () => {
```
Comment (mcmire) [inline src/keyring/KeyringController.test.ts:671]: We should probably make the name of this test match the current behavior in the meantime, even if it's wrong.

File: src/keyring/KeyringController.test.ts [hunk 7]
```diff
@@ -104,11 +106,13 @@ describe('KeyringController', () => {
...
   });
 
-  it('should create new vault and keychain', async () => {
+  it('should create new vault, mnemonic and keychain', async () => {
     const currentState = await keyringController.createNewVaultAndKeychain(
       password,
     );
     expect(initialState).not.toBe(currentState);
+    const currentSeedWord = await keyringController.exportSeedPhrase(password);
+    expect(currentSeedWord).toBeDefined();
```
Comment (mcmire) [inline src/keyring/KeyringController.test.ts:115]: Is there any way for us to stub whatever is responsible for generating the seed phrase so we can be more exact with this assertion?

File: src/keyring/KeyringController.test.ts [hunk 8]
```diff
@@ -508,6 +699,11 @@ describe('KeyringController', () => {
     expect(listenerUnlock.called).toBe(true);
   });
 
+  it('should return current seedphrase', async () => {
+    const seedPhrase = await keyringController.verifySeedPhrase();
+    expect(seedPhrase).toBeDefined();
```
Comment (mcmire) [inline src/keyring/KeyringController.test.ts:704]: Is there any way for us to stub whatever is responsible for generating the seed phrase so we can be more exact with this assertion?

File: src/keyring/KeyringController.test.ts [hunk 9]
```diff
@@ -449,8 +563,42 @@ describe('KeyringController', () => {
...
+      ];
+      await keyringController.signTypedMessage(
+        { data: typedMsgParams, from: '' },
+        SignTypedDataVersion.V1,
+      );
+    } catch (e) {
+      error = e;
+    }
+
+    expect((error as ErrorMessageObject).message).toContain(
```
Comment (mcmire) [inline src/keyring/KeyringController.test.ts:599]: Same for this test — use `expect(...).rejects.toThrow(...)`?

File: src/keyring/KeyringController.test.ts (hunk 2 above)
Comment (mcmire) [reply src/keyring/KeyringController.test.ts:650]: How can the check be improved? Do you mind filing a new issue with more information?

File: src/keyring/KeyringController.test.ts (hunk 6 above)
Comment (mcmire) [reply src/keyring/KeyringController.test.ts:671]: Do you mind filing a new issue that notes this?

File: src/keyring/KeyringController.ts [hunk 10]
```diff
@@ -380,6 +384,11 @@ export class KeyringController extends BaseController<
     return this.fullUpdate();
   }
 
+  /**
```
Comment (mcmire) [inline src/keyring/KeyringController.ts:387]: I am not certain but do you mind filing a new issue with this question? That way we can start a discussion about it there.

**Review by mcmire — COMMENTED** [review]
Hey @jpuri, sorry for the delay in reviewing this. I need to take a closer look but for now I've just added comments for style-related things. @Gudahtt You might want to take a look at this too. 

Just a note about the TODOs. I am not a big fan of adding TODOs to code because I feel like it creates noise when reading through the code. Would you mind filing issues around problems you see? You can copy/paste the comments you've left if you like. That should allow us to have discussions about these things in a dedicated place and give visibility for @Gudahtt as he works through the keyring stuff.

File: src/keyring/KeyringController.test.ts (hunk 7 above)
Comment (adonesky1) [reply src/keyring/KeyringController.test.ts:115]: This would be stubbing `generateRandomMnemonic` on  `HDKeyring` from `@metamask/eth-hd-keyring` just fyi @jpuri

File: src/keyring/KeyringController.test.ts [hunk 11]
```diff
@@ -119,7 +123,29 @@ describe('KeyringController', () => {
     expect(initialState).not.toBe(currentState);
   });
 
+  it('should restore same vault if old seedWord is used', async () => {
+    const currentSeedWord = await keyringController.exportSeedPhrase(password);
+    const currentState = await keyringController.createNewVaultAndRestore(
+      password,
+      currentSeedWord,
+    );
+    expect(initialState).toStrictEqual(currentState);
```
Comment (adonesky1) [inline src/keyring/KeyringController.test.ts:132]: 👍

File: src/keyring/KeyringController.test.ts [hunk 12]
```diff
@@ -144,6 +170,14 @@ describe('KeyringController', () => {
...
   });
 
+  it('should not export account if address is not provided', async () => {
+    await expect(
+      async () => await keyringController.exportAccount(password, ''),
+    ).rejects.toThrow(
+      'No keyring found for the requested account. Error info: The address passed in is invalid/empty; There are keyrings, but none match the address;',
+    );
+  });
+
```
Comment (adonesky1) [inline src/keyring/KeyringController.test.ts:180]: maybe another test(s) `should not export account if password is not provided` and/or `should not export account if password provided is incorrect`>

File: src/keyring/KeyringController.test.ts [hunk 13]
```diff
@@ -119,7 +123,29 @@ describe('KeyringController', () => {
...
+    expect(initialState).toStrictEqual(currentState);
+  });
+
+  it('should throw error if creating new vault and restore without password', async () => {
+    await expect(async () => {
+      await keyringController.createNewVaultAndRestore('', seedWords);
+    }).rejects.toThrow('Invalid password');
+  });
+
+  it('should throw error if creating new vault and restore without seedWord', async () => {
```
[the remark this answers is not in the document]
Comment (jpuri) [reply src/keyring/KeyringController.test.ts:141]: Fixed 👍

File: src/keyring/KeyringController.test.ts (hunk 12 above)
Comment (jpuri) [reply src/keyring/KeyringController.test.ts:182]: yep added

File: src/keyring/KeyringController.test.ts (hunk 3 above)
Comment (jpuri) [reply src/keyring/KeyringController.test.ts:659]: Created issue https://github.com/MetaMask/controllers/issues/798 to look into it.

File: src/keyring/KeyringController.test.ts (hunk 5 above)
Comment (jpuri) [reply src/keyring/KeyringController.test.ts]: Fixed

File: src/keyring/KeyringController.test.ts (hunk 6 above)
Comment (jpuri) [reply src/keyring/KeyringController.test.ts:663]: Yep added [issue](https://github.com/MetaMask/controllers/issues/798)

File: src/keyring/KeyringController.test.ts (hunk 7 above)
Comment (jpuri) [reply src/keyring/KeyringController.test.ts]: `@metamask/eth-hd-keyring` is a transitive dependency and I found it hard to mock it.

File: src/keyring/KeyringController.test.ts (hunk 8 above)
Comment (jpuri) [reply src/keyring/KeyringController.test.ts:696]: same as above https://github.com/MetaMask/controllers/pull/693#discussion_r857401548

File: src/keyring/KeyringController.test.ts (hunk 9 above)
Comment (jpuri) [reply src/keyring/KeyringController.test.ts]: Fixed

File: src/keyring/KeyringController.test.ts (hunk 2 above)
Comment (jpuri) [reply src/keyring/KeyringController.test.ts:642]: yep done https://github.com/MetaMask/controllers/issues/800

File: src/keyring/KeyringController.test.ts (hunk 1 above)
Comment (jpuri) [reply src/keyring/KeyringController.test.ts:368]: Created issue https://github.com/MetaMask/controllers/issues/799

File: src/keyring/KeyringController.ts [hunk 14]
```diff
@@ -366,6 +370,11 @@ export class KeyringController extends BaseController<
     return this.fullUpdate();
   }
```
[the remark this answers is not in the document]
Comment (jpuri) [reply src/keyring/KeyringController.ts]: https://github.com/MetaMask/controllers/issues/801

File: src/keyring/KeyringController.ts (hunk 10 above)
Comment (jpuri) [reply src/keyring/KeyringController.ts]: https://github.com/MetaMask/controllers/issues/801

**Comment by jpuri** [comment]
Test coverage in KeyringController after PR:

Screenshot 2022-04-25 at 3 43 47 PM

File: src/keyring/KeyringController.test.ts (hunk 1 above)
Comment (adonesky1) [reply src/keyring/KeyringController.test.ts:378]: 🤔

**Review by adonesky1 — DISMISSED** [review]
LGTM!

File: src/keyring/KeyringController.test.ts [hunk 15]
```diff
@@ -230,6 +288,26 @@ describe('KeyringController', () => {
...
+      keyringController.importAccountWithStrategy(AccountImportStrategy.json, [
+        input,
+        '',
+      ]),
+    ).rejects.toThrow('Key derivation failed - possibly wrong passphrase');
+  });
+
+  /**
+   * If there is only HD Key Tree keyring with 1 account and removeAccount is called passing that account
+   * It deletes keyring object also from state - not sure if this is correct behavior.
```
Comment (Gudahtt) [inline src/keyring/KeyringController.test.ts:302]: Interesting. Right now we don't make it possible to delete these accounts, so we never encountered this. I agree that it seems like the wrong thing to do.

**Review by Gudahtt — DISMISSED** [review]
LGTM! Still lots of room for improvement but this seems like a clear step forward.

File: src/keyring/KeyringController.test.ts [hunk 16]
```diff
@@ -469,11 +628,56 @@ describe('KeyringController', () => {
...
+  /**
+   * Task added to improve check for valid transaction in signTransaction method
+   * https://github.com/MetaMask/controllers/issues/800
+   */
+  it('should not sign transaction if transaction is not valid', async () => {
+    await expect(async () => {
+      const account = initialState.keyrings[0].accounts[0];
+      const signedTx = await keyringController.signTransaction({}, account);
+      expect(signedTx.v).not.toBeUndefined();
+      expect(signedTx).not.toBe('');
```
Comment (mcmire) [inline src/keyring/KeyringController.test.ts:663]: [nit] I think you can drop these `expect`s, yeah? Because it should error before this point.

**Review by mcmire — APPROVED** [review]
Noticed one small thing, but it is minor. Looks good to me regardless.

**Comment by jpuri** [comment]
I have done a small update, can you plz re-approve @adonesky1 , @mcmire , @Gudahtt

File: src/keyring/KeyringController.test.ts (hunk 16 above)
Comment (jpuri) [reply src/keyring/KeyringController.test.ts]: Good point fixed.

**Review by Gudahtt — APPROVED** [review]

File: src/keyring/KeyringController.test.ts [hunk 17]
```diff
@@ -445,6 +583,27 @@ describe('KeyringController', () => {
...
+        type: 'uint32',
+        value: '1337',
+      },
+    ];
+    await expect(
+      keyringController.signTypedMessage(
+        { data: typedMsgParams, from: '' },
+        SignTypedDataVersion.V1,
+      ),
+    ).rejects.toThrow(/^Keyring Controller signTypedMessage:/u);
```
[the remark this answers is not in the document]
Comment (jpuri) [reply src/keyring/KeyringController.test.ts:604]: This is only matching string. Actual error message is long
